pl_bolts/models/rl/ppo_model.py

```diff
@@ -0,0 +1,375 @@
+            self.actor = ActorCategorical(actor_mlp)
+        else:
+            raise NotImplementedError(
+                'Env action space should be of type Box (continous) or Discrete (categorical). '
+                f'Got type: {type(self.env.action_space)}'
+            )
+
+        self.batch_states = []
+        self.batch_actions = []
+        self.batch_adv = []
+        self.batch_qvals = []
+        self.batch_logp = []
+
+        self.ep_rewards = []
+        self.ep_values = []
+        self.epoch_rewards = []
+
+        self.episode_step = 0
+        self.avg_ep_reward = 0
+        self.avg_ep_len = 0
+        self.avg_reward = 0
+
+        self.state = torch.FloatTensor(self.env.reset())
+
+    def forward(self, x: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        """
+        Passes in a state x through the network and returns the policy and a sampled action
+        Args:
+            x: environment state
+        Returns:
+            Tuple of policy and action
+        """
+        pi, action = self.actor(x)
+        value = self.critic(x)
+
+        return pi, action, value
+
+    def discount_rewards(self, rewards: List[float], discount: float) -> List[float]:
+        """Calculate the discounted rewards of all rewards in list
+
+        Args:
+            rewards: list of rewards/advantages
+
+        Returns:
+            list of discounted rewards/advantages
+        """
+        assert isinstance(rewards[0], float)
+
+        cumul_reward = []
+        sum_r = 0.0
+
+        for r in reversed(rewards):
+            sum_r = (sum_r * discount) + r
+            cumul_reward.append(sum_r)
+
+        return list(reversed(cumul_reward))
+
+    def calc_advantage(self, rewards: List[float], values: List[float], last_value: float) -> List[float]:
+        """Calculate the advantage given rewards, state values, and the last value of episode
+
+        Args:
+            rewards: list of episode rewards
+            values: list of state values from critic
+            last_value: value of last state of episode
+
+        Returns:
+            list of advantages
+        """
+        rews = rewards + [last_value]
+        vals = values + [last_value]
+        # GAE
+        delta = [rews[i] + self.gamma * vals[i + 1] - vals[i] for i in range(len(rews) - 1)]
+        adv = self.discount_rewards(delta, self.gamma * self.lam)
+
+        return adv
+
+    def generate_trajectory_samples(self) -> Tuple[List[torch.Tensor], List[torch.Tensor], List[torch.Tensor]]:
+        """
+        Contains the logic for generating trajectory data to train policy and value network
+
+        Yield:
+           Tuple of Lists containing tensors for states, actions, log probs, qvals and advantage
+        """
+
+        for step in range(self.steps_per_epoch):
+            self.state = self.state.to(device=self.device)
+
+            with torch.no_grad():
```

Comment:
codecov complain that all the following is not tested, mind add test for it...

Reply:
added :)